Add MinDomains API to TopologySpreadConstraints field

sanposhiho opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Adds MinDomains API to TopologySpreadConstraints field

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of #105291

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
TopologySpreadConstraints includes minDomains field to limit the minimum number of topology domains.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3022
```

sanposhiho commented:
/cc @alculquicondor

alculquicondor commented:
/label api-review

alculquicondor commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3390,6 +3390,14 @@ type TopologySpreadConstraint struct {
+	// scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same or more number of matching Pods as `maxSkew`.
+	// Default value is 0. When value is different than 0, WhenUnsatisfiable must be DoNotSchedule.
+	// +optional
+	MinDomains *int32
```
you are missing the json tag. The protobuff one can be autogenerated running `make update`.

sanposhiho replied:
Oops. Thanks.

alculquicondor reviewed (commented):
You are missing:
- defaults in https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/defaults.go
- a strategy to drop the field when the feature is disabled https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/strategy.go

Because of the later, the feature gate actually affects the kube-apiserver. Please update the KEP to reflect that.

denkensk commented on api/openapi-spec/swagger.json:
```diff
@@ -10539,6 +10539,11 @@
           "type": "string"
+        },
+        "minDomains": {
+          "description": "MinDomains indicates a minimum number of domains. When the number of domains with matching topology keys is less than `minDomains`, Pod Topology Spread treats\"global minimu\" as 0. As a result, when the number of domains is less than `minDomains`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same or more number of matching Pods as `max ...
```
```suggestion
          "description": "MinDomains indicates a minimum number of domains. When the number of domains with matching topology keys is less than `minDomains`, Pod Topology Spread treats\"global minimum\" as 0. As a result, when the number of domains is less than `minDomains`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same or more number of matching Pods as `maxSkew`. Default value is 0. When value is different from 0, WhenUnsatisfiable must be DoNotSchedule.",
```

denkensk commented on pkg/scheduler/framework/plugins/podtopologyspread/plugin.go:
```diff
@@ -95,6 +97,7 @@ func New(plArgs runtime.Object, h framework.Handle) (framework.Plugin, error) {
 		parallelizer:       h.Parallelizer(),
 		sharedLister:       h.SnapshotSharedLister(),
 		defaultConstraints: args.DefaultConstraints,
+		feature:            fts,
```
Better to add a field `enableMinDomainsInPodTopologySpread` here

denkensk commented on pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:
```diff
@@ -518,7 +523,7 @@ func TestPreFilterState(t *testing.T) {
 				DefaultingType:     config.ListDefaulting,
 			}
-			p := plugintesting.SetupPluginWithInformers(ctx, t, New, args, cache.NewSnapshot(tt.existingPods, tt.nodes), tt.objs)
+			p := plugintesting.SetupPluginWithInformers(ctx, t, frameworkruntime.FactoryAdapter(plfeature.Features{}, New), args, cache.NewSnapshot(tt.existingPods, tt.nodes), tt.objs)
```
You can create a variable that can be used here to make code cleaner. 
You can refer to this https://github.com/kubernetes/kubernetes/blob/45b6af6b71db0c5c61596add8b3f2c67dd071727/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go#L98

denkensk commented on pkg/scheduler/framework/plugins/podtopologyspread/common.go:
```diff
@@ -66,19 +68,27 @@ func nodeLabelsMatchSpreadConstraints(nodeLabels map[string]string, constraints
+				// The default value of MinDomains is 0.
+				tsc.MinDomains = 0
+				if c.MinDomains == nil {
+					tsc.MinDomains = *c.MinDomains
```
Bug here?  c.MinDomains is nil.  It is better to add UT for this.
`c.MinDomains != nil` ?

sanposhiho replied:
Thanks for nice catching... 🙏 
Exactly, I added the unit test for this.

sanposhiho commented:
@denkensk @alculquicondor 
Thanks for reviewing. I've updated as your suggestions.

--

@alculquicondor 
> a strategy to drop the field when the feature is disabled
https://github.com/kubernetes/kubernetes/pull/107674#pullrequestreview-858596259

In terms of this, aren't [the changes](https://github.com/kubernetes/kubernetes/pull/107674#pullrequestreview-858596259) in pkg/api/pod/util.go `DropDisabledPodFields` enough?
It seems `PrepareForCreate` and `PrepareForUpdate` funcs call this `podutil.DropDisabledPodFields` to drop the field.
https://github.com/kubernetes/kubernetes/blob/2a2b294c999894a11f64cde5de7280bed6171b30/pkg/registry/core/pod/strategy.go#L81-L101
Or do I need to add another change to /pkg/registry/core/pod/strategy.go?

> Because of the later, the feature gate actually affects the kube-apiserver. Please update the KEP to reflect that.

We already have added kube-apiserver to "Components depending on the feature gate" as @\Huang-Wei  's suggestion. 
https://github.com/kubernetes/enhancements/pull/3030#discussion_r758603750
Let me know if I need to fix another part in KEP.

alculquicondor commented on pkg/api/pod/util.go:
```diff
@@ -591,6 +591,35 @@ func dropDisabledFields(
+	if !utilfeature.DefaultFeatureGate.Enabled(features.MinDomainsInPodTopologySpread) &&
+		!minDomainsInUse(oldPodSpec) &&
+		podSpec != nil {
+		for i := range podSpec.TopologySpreadConstraints {
```
It's more appropriate to check the usage in each constraint independently.

Otherwise, if there is one constraint with the field used, one could add the field in other constraint, even if the feature is disabled.

The purpose of the strategy is that, if the feature is disabled, you only allow existing fields to remain set, but not to add new usages.

liggitt replied:
correlating the constraints between the old and new objects is probably more trouble than it's worth... an object-wide permit/clear decision is probably fine

sanposhiho replied:
I updated the implementation as @alculquicondor suggestion. 

@liggitt if you have any concerns about the implementation, let me know.

alculquicondor replied:
Please follow @liggitt's advice instead.

sanposhiho replied:
Sure.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5606,6 +5606,7 @@ type TopologySpreadConstraint struct {
 	// between the number of matching pods in the target topology and the global minimum.
+	// The global minimum indicates the minimum number of pods in an eligible domain.
 	// For example, in a 3-zone cluster, MaxSkew is set to 1, and pods with the same
 	// labelSelector spread as 1/1/0:
```
say that, in this example, the global minimum is zero. Although it would be more useful to have the example as 2/2/1, such that global minimum is not zero, to distinguish from the `minDomains` example.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5625,6 +5626,8 @@ type TopologySpreadConstraint struct {
 	// and identical values are considered to be in the same topology.
 	// We consider each <key, value> as a "bucket", and try to put balanced number
 	// of pods into each bucket.
+	// A domain indicates a particular instance of a topology.
```
```suggestion
	// We define a domain as a particular instance of a topology.
```

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5625,6 +5626,8 @@ type TopologySpreadConstraint struct {
 	// We consider each <key, value> as a "bucket", and try to put balanced number
 	// of pods into each bucket.
+	// A domain indicates a particular instance of a topology.
+	// e.g. If TopologyKey is "kubernetes.io/hostname", a domain of that topology is each Node.
```
```suggestion
	// e.g. If TopologyKey is "kubernetes.io/hostname", each Node is a domain of that topology.
```

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5654,6 +5657,25 @@ type TopologySpreadConstraint struct {
+	// When the number of domains with matching topology keys is less than `minDomains`,
+	// Pod Topology Spread treats "global minimum" as 0.
+	// As a result, when the number of domains is less than `minDomains`,
+	// scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same or more number of matching Pods as `maxSkew`.
```
Also missing that values of minDomains must be greater than zero.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5654,6 +5657,25 @@ type TopologySpreadConstraint struct {
+	// As a result, when the number of domains is less than `minDomains`,
+	// scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same or more number of matching Pods as `maxSkew`.
+	// Default value is 1. When value is different from 1, WhenUnsatisfiable must be DoNotSchedule.
+	// This is an alpha field and requires enabling MinDomainsInPodTopologySpread feature.
```
It's better to put this at the end of the comment, after an empty `//` line, for better visibility.

sanposhiho replied:
Sorry, I misunderstood the comment.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5654,6 +5657,25 @@ type TopologySpreadConstraint struct {
+	// +-------+-------+-------+
+	// The number of domains is less than 5(MinDomains), so "global minimum" is treated as 0.
+	// In this situation, new pod with the same labelSelector cannot be scheduled,
+	// because computed skew will be 3(3 - 0) if new Pod is scheduled to a domain, and it will violate MinDomains.
```
```suggestion
	// because computed skew will be 3(3 - 0) if new Pod is scheduled to any of the three zones, and it will violate MinDomains.
```

alculquicondor commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -180,6 +180,15 @@ func SetDefaults_Pod(obj *v1.Pod) {
+		for i := range obj.Spec.TopologySpreadConstraints {
+			if obj.Spec.TopologySpreadConstraints[i].MinDomains == nil {
+				defaultMinDomains := v1.DefaultMinDomains
+				obj.Spec.TopologySpreadConstraints[i].MinDomains = &defaultMinDomains
```
use `pointer.Int32`

alculquicondor reviewed (commented):
I'm only reviewing the API for now. @denkensk let me know when you think the implementation looks good.

fedebongio commented:
/triage accepted
/remove-sig api-machinery

denkensk commented on pkg/scheduler/framework/plugins/podtopologyspread/common.go:
```diff
@@ -66,19 +67,27 @@ func nodeLabelsMatchSpreadConstraints(nodeLabels map[string]string, constraints
-			})
+			}
+			if enableMinDomainsInPodTopologySpread {
+				defaultMinDomains := v1.DefaultMinDomains
+				tsc.MinDomains = &defaultMinDomains
+				if c.MinDomains != nil {
+					tsc.MinDomains = c.MinDomains
+				}
```
```suggestion

				if c.MinDomains != nil {
					tsc.MinDomains = c.MinDomains
				} else {
					defaultMinDomains := v1.DefaultMinDomains
				        tsc.MinDomains = &defaultMinDomains
				}
```

denkensk commented on pkg/scheduler/framework/plugins/podtopologyspread/common_test.go:
```diff
@@ -0,0 +1,138 @@
+					{
+						WhenUnsatisfiable: v1.DoNotSchedule,
+						LabelSelector:     labelSelector,
+						MinDomains:        ptrInt32(2),
```
Test if the default value takes effect when it is not set？

sanposhiho commented:
@alculquicondor @denkensk 
Thanks for reviewing. 🙏 I updated as your suggestions.

sanposhiho commented:
/retest

sanposhiho commented:
/retest

alculquicondor commented:
I need to prioritize reviewing KEPs, so I might only take a look at this next week.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5654,6 +5659,29 @@ type TopologySpreadConstraint struct {
+	// The number of domains is less than 5(MinDomains), so "global minimum" is treated as 0.
+	// In this situation, new pod with the same labelSelector cannot be scheduled,
+	// because computed skew will be 3(3 - 0) if new Pod is scheduled to any of the three zones,
+	// and it will violate MinDomains.
```
```suggestion
	// it will violate MinDomains.
```

alculquicondor commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3337,9 +3337,16 @@ const (
 
+const (
+	// DefaultMinDomains indicates the default value for MinDomains field in TopologySpreadConstraint.
+	DefaultMinDomains int32 = 1
```
Remove this constant.

sanposhiho commented:
@alculquicondor 
Thanks for reviewing. Updated as your suggestions.

sanposhiho commented:
@denkensk @alculquicondor Could you please take another look at this PR as well? 🙏

alculquicondor commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -23,6 +23,7 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/kubernetes/pkg/util/parsers"
+	"k8s.io/utils/pointer"
```
this is already imported one line below :)

Although, if there is no conflicts, it's better to to set an alias. In other words, you can keep line 26, but remove 27.

alculquicondor commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -18693,6 +18693,22 @@ func TestValidateTopologySpreadConstraints(t *testing.T) {
+					MaxSkew:           1,
+					TopologyKey:       "k8s.io/zone",
+					WhenUnsatisfiable: core.DoNotSchedule,
+					MinDomains: func() *int32 {
```
use pointer package

alculquicondor commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -18693,6 +18693,22 @@ func TestValidateTopologySpreadConstraints(t *testing.T) {
 			errfield: "maxSkew",
 		},
+		{
+			name: "invalid minDomains",
```
Also test a valid minDomains :)

sanposhiho replied:
Sure. Added to "all required fields ok" test case.

alculquicondor commented on pkg/scheduler/framework/plugins/podtopologyspread/common.go:
```diff
@@ -36,13 +37,14 @@ type topologySpreadConstraint struct {
 	MaxSkew     int32
 	TopologyKey string
 	Selector    labels.Selector
+	MinDomains  *int32
```
This can just be `int32`

alculquicondor commented on pkg/scheduler/framework/plugins/podtopologyspread/common.go:
```diff
@@ -66,19 +68,28 @@ func nodeLabelsMatchSpreadConstraints(nodeLabels map[string]string, constraints
+			tsc := topologySpreadConstraint{
 				MaxSkew:     c.MaxSkew,
 				TopologyKey: c.TopologyKey,
 				Selector:    selector,
```
you can just set it to 1 when the feature is disabled.

alculquicondor commented on pkg/scheduler/framework/plugins/podtopologyspread/common_test.go:
```diff
@@ -0,0 +1,178 @@
+	"k8s.io/utils/pointer"
+)
+
+func Test_filterTopologySpreadConstraints(t *testing.T) {
```
We generally don't add unit tests for non-exported functions, unless it has a very complicated logic and it's hard to get good coverage from the exported functions.

I think we should be good without one in this case.

sanposhiho replied:
Got it. 👍

alculquicondor commented on pkg/scheduler/framework/plugins/podtopologyspread/common_test.go:
```diff
@@ -0,0 +1,178 @@
+				t.Errorf("filterTopologySpreadConstraints() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
```
for future reference: always prefer cmp.Diff to reflect.DeepEqual

alculquicondor commented on pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:
```diff
@@ -203,7 +203,7 @@ func (pl *PodTopologySpread) calPreFilterState(pod *v1.Pod) (*preFilterState, er
 		// We have feature gating in APIServer to strip the spec
 		// so don't need to re-check feature gate, just check length of Constraints.
-		constraints, err = filterTopologySpreadConstraints(pod.Spec.TopologySpreadConstraints, v1.DoNotSchedule)
+		constraints, err = filterTopologySpreadConstraints(pod.Spec.TopologySpreadConstraints, v1.DoNotSchedule, pl.enableMinDomainsInPodTopologySpread)
```
Where is the code that uses the MinDomains?

sanposhiho replied:
I was thinking of implementing the core logic of MinDomains in the next PR, but should I include it in this PR?

alculquicondor replied:
Usually we either split only API changes and adding the feature gate in one PR and another one for using the field. Or you might just put everything together, if the changes are small enough.

But since this is already reviewed, it's fine. This can't merge until the implementation is ready though. Better do it in a separate PR at this point.

sanposhiho replied:
Okay. I'll prepare the PR for the implementation.

sanposhiho replied:
Open #108362

liggitt commented on staging/src/k8s.io/api/testdata/HEAD/apps.v1.DaemonSet.json:
```diff
@@ -1575,40 +1575,41 @@
               ]
-            }
+            },
+            "minDomains": 1680820399
```
FYI, I'm reworking how these fixtures are generated in https://github.com/kubernetes/kubernetes/pull/108219 so that a tiny change like this PR would not ripple into huge changes in these fixtures. After my fix, this PR would result in an update like https://github.com/liggitt/kubernetes/commit/a78331610dd94ef9bac5f8f4ab8a2d0b6a820851

liggitt replied:
https://github.com/kubernetes/kubernetes/pull/108219 is now merged

The easiest way to rebase on master and rerun hack/update-generated-api-compatibility-data.sh would probably be something like:

```
# Revert changes to HEAD fixture files
git checkout d5263fe -- staging/src/k8s.io/api/testdata

# Squash to a single commit
git reset --soft d5263fe
git commit -m "Add MinDomains API to TopologySpreadConstraints field"

# Rebase on master (assumes your remote for Kubernetes is named "upstream")
git fetch upstream
git rebase upstream/master

# Regenerate fixture files
hack/update-generated-api-compatibility-data.sh
git add staging/src/k8s.io/api/testdata
git commit -m "Generated files"
```

sanposhiho replied:
@liggitt Thanks for sharing. I'll try that.

sanposhiho commented:
@alculquicondor 
Thanks. I've addressed all your reviews. 
And if you want this PR to include the core logic of MinDomains, I'll add that. https://github.com/kubernetes/kubernetes/pull/107674#discussion_r810568145

sanposhiho committed:
Add MinDomains API to TopologySpreadConstraints field

alculquicondor commented on pkg/api/pod/util_test.go:
```diff
@@ -1519,6 +1519,193 @@ func TestHaveSameExpandedDNSConfig(t *testing.T) {
+			podSpec: &api.PodSpec{
+				TopologySpreadConstraints: []api.TopologySpreadConstraint{
+					{
+						MinDomains: func() *int32 {
```
Still pending

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5624,6 +5627,10 @@ type TopologySpreadConstraint struct {
 	// We consider each <key, value> as a "bucket", and try to put balanced number
 	// of pods into each bucket.
+	// We define a domain as a particular instance of a topology.
+	// Also, we define eligible domain as a domain that matches the node selector.
```
```suggestion
	// Also, we define an eligible domain as a domain whose nodes match the node selector.
```

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5624,6 +5627,10 @@ type TopologySpreadConstraint struct {
+	// We define a domain as a particular instance of a topology.
+	// Also, we define eligible domain as a domain that matches the node selector.
+	// e.g. If TopologyKey is "kubernetes.io/hostname", each Node is a domain of that topology.
+	// And, if TopologyKey is "topology.kubernetes.io/zone", a domain of that topology is formed by Nodes in a zone.
```
```suggestion
	// And, if TopologyKey is "topology.kubernetes.io/zone", each zone is a domain of that topology.
```

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5654,6 +5659,30 @@ type TopologySpreadConstraint struct {
 	// +optional
 	LabelSelector *metav1.LabelSelector
+	// MinDomains indicates a minimum number of domains.
+	// When the number of domains with matching topology keys is less than `minDomains`,
```
Still not addressed

alculquicondor commented on pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:
```diff
@@ -23,13 +23,16 @@ import (
 	"testing"
 
 	"github.com/google/go-cmp/cmp"
+
```
remove this line

sanposhiho commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6518,6 +6518,9 @@ func validateTopologySpreadConstraints(constraints []core.TopologySpreadConstrai
 		if err := ValidateMaxSkew(subFldPath.Child("maxSkew"), constraint.MaxSkew); err != nil {
 			allErrs = append(allErrs, err)
 		}
+		if err := validateMinDomains(subFldPath.Child("minDomains"), constraint.MinDomains); err != nil {
```
Thanks.

sanposhiho commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6541,6 +6544,21 @@ func ValidateMaxSkew(fldPath *field.Path, maxSkew int32) *field.Error {
+		}
+
+		// When MinDomains is different from 1, WhenUnsatisfiable must be DoNotSchedule.
+		if action != core.DoNotSchedule && *minDomains > 0 {
```
Thanks.

alculquicondor commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6541,6 +6544,21 @@ func ValidateMaxSkew(fldPath *field.Path, maxSkew int32) *field.Error {
 }
 
+// validateMinDomains tests that the argument is a valid MinDomains.
+func validateMinDomains(fldPath *field.Path, minDomains *int32, action core.UnsatisfiableConstraintAction) *field.Error {
```
you should return all errors `[]*field.Error`

alculquicondor commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6541,6 +6544,21 @@ func ValidateMaxSkew(fldPath *field.Path, maxSkew int32) *field.Error {
+
+		// When MinDomains is different from 1, WhenUnsatisfiable must be DoNotSchedule.
+		if action != core.DoNotSchedule && *minDomains != 1 {
+			return field.Invalid(fldPath, minDomains, "cannot use MinDomains with "+string(action))
```
```suggestion
			return field.Invalid(fldPath, minDomains, "cannot use MinDomains with whenUnsatifiable="+string(action))
```

alculquicondor reviewed (commented):
Please squash so it's ready to be merged once the implementation is ready

You can choose to keep the generated files in a separate commit for api-reviewers convenience.

Huang-Wei commented on pkg/apis/core/types.go:
```diff
@@ -5653,6 +5660,29 @@ type TopologySpreadConstraint struct {
+	//
+	// This is an alpha field and requires enabling MinDomainsInPodTopologySpread feature gate.
+	// +optional
+	MinDomains *int32
```
This (internal version) should be a concrete value

```suggestion
	MinDomains int32
```

liggitt replied:
No, it should be a pointer. If it is a non-pointer, we cannot round trip through internal without populating the field value, which we don't want to do before the field graduates.

Huang-Wei replied:
Good to know that. I've followed the "concrete value for versioned type and pointer for unversioned types" for a while..

So when it graduates to GA, is it optional or mandatory to make it a concret value? Also, does it relate with optional/required semantics?

liggitt replied:
> Good to know that. I've followed the "concrete value for versioned type and pointer for unversioned types" for a while..

other way around, right? pointer for versioned type and non-pointer for unversioned types?

> So when it graduates to GA, is it optional or mandatory to make it a concret value?

if you still need to distinguish between set/unset (for defaulting or validation purposes), it should still be a pointer.

Huang-Wei replied:
> other way around, right? pointer for versioned type and non-pointer for unversioned types?

yes, typo...

> if you still need to distinguish between set/unset (for defaulting or validation purposes)

It sounds o be applicable to the versioned type? I have no question on the versioned type - it should be always a pointer as we need default unset to 1. My confusion was on the unversioned (internal) type: should it be non-pointer since the beginning, or updated to non-pointer when it graduates?

sanposhiho commented:
@liggitt 
Sorry for bothering you, do you have any idea why there is an error in the generated file? 
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/107674/pull-kubernetes-verify/1498976880995340288

I've already run `make update`.

Huang-Wei commented:
The latest changes look good.

alculquicondor commented:
Can you squash?

kerthcet commented on pkg/api/pod/util_test.go:
```diff
@@ -1519,6 +1520,174 @@ func TestHaveSameExpandedDNSConfig(t *testing.T) {
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.MinDomainsInPodTopologySpread, tc.enabled)()
+
```
nit: remove this blank line.

kerthcet commented:
One nit, besides this also looks good to me then.

sanposhiho commented:
Thanks everyone for reviewing. squashed.

alculquicondor reviewed (commented):
I always manage to find something last minute :sweat_smile: 
But they are on the nit side. I'm fine if you amend the commits this time.

sanposhiho commented:
@alculquicondor Thanks. Updated as your suggestions.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6541,6 +6544,22 @@ func ValidateMaxSkew(fldPath *field.Path, maxSkew int32) *field.Error {
+
+		// When MinDomains is different from 1, whenUnsatisfiable must be DoNotSchedule.
+		if action != core.DoNotSchedule && *minDomains != 1 {
+			allErrs = append(allErrs, field.Invalid(fldPath, minDomains, fmt.Sprintf("cannot use MinDomains with whenUnsatisfiable=%s", string(action))))
```
use the serialized API name and tell them what whenUnsatisfiable must be:
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath, minDomains, fmt.Sprintf("can only use minDomains if whenUnsatisfiable=%s, not %s", string(core.DoNotSchedule), string(action))))
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6518,6 +6518,9 @@ func validateTopologySpreadConstraints(constraints []core.TopologySpreadConstrai
 		if err := ValidateMaxSkew(subFldPath.Child("maxSkew"), constraint.MaxSkew); err != nil {
 			allErrs = append(allErrs, err)
 		}
+		if errs := validateMinDomains(subFldPath.Child("minDomains"), constraint.MinDomains, constraint.WhenUnsatisfiable); len(errs) != 0 {
```
simplify to:
```go
allErrs = append(allErrs, validateMinDomains(subFldPath.Child("minDomains"), constraint.MinDomains, constraint.WhenUnsatisfiable)...)
```

alculquicondor commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -180,6 +179,15 @@ func SetDefaults_Pod(obj *v1.Pod) {
 		obj.Spec.EnableServiceLinks = &enableServiceLinks
 	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.MinDomainsInPodTopologySpread) {
```
I disagree that this would be a problem, but I'll discuss in the new PR.

alculquicondor replied:
Clarified offline that the concern is specific to PodSpec.

@sanposhiho the conclusion is that you should delete the defaulting code and make sure that kube-scheduler understands `nil`.

liggitt replied:
(and modify the API doc to not refer to a default value, and instead describe what it means if the field is nil)

sanposhiho replied:
Sure. 👍

sanposhiho replied:
Fixed. Let me know if I miss something for that 🙏  
https://github.com/kubernetes/kubernetes/pull/107674/commits/5adad2316ac5f3739bcb70a20c2d6ee6092f568a

kerthcet commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6541,6 +6542,22 @@ func ValidateMaxSkew(fldPath *field.Path, maxSkew int32) *field.Error {
+		allErrs = append(allErrs, field.Invalid(fldPath, minDomains, isNotPositiveErrorMsg))
+	}
+	// When MinDomains is non-nil, whenUnsatisfiable must be DoNotSchedule.
+	if action != core.DoNotSchedule {
```
In `SetDefaults_Pod`, `MinDomains` will default to 1. We should check both `action` and `MinDomains` value here, right?

alculquicondor replied:
We just changed the default to be `nil` (that is, we don't default anything). This validation is correctly matching the defined API.

sanposhiho replied:
Ah, okay. I forgot to remove the code for MinDomains default value in `SetDefaults_Pod`. Thanks.

alculquicondor commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -18739,19 +18782,75 @@ func TestValidateTopologySpreadConstraints(t *testing.T) {
+				} else if !strings.HasSuffix(errs[0].Field, "."+tc.errfield) {
+					t.Errorf("[%d: %q] expected error on field %q, got %q", i, tc.name, tc.errfield, errs[0].Field)
+				}
+			}
```
It doesn't look like we are checking that the errors match in all cases.

sanposhiho replied:
I change `TestValidateTopologySpreadConstraints` to simply compare the result with the expected value.

alculquicondor commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -18739,19 +18782,75 @@ func TestValidateTopologySpreadConstraints(t *testing.T) {
-				t.Errorf("[%d: %q] expected error type %v, got %v", i, tc.name, tc.errtype, errs[0].Type)
-			} else if !strings.HasSuffix(errs[0].Field, "."+tc.errfield) {
-				t.Errorf("[%d: %q] expected error on field %q, got %q", i, tc.name, tc.errfield, errs[0].Field)
+func Test_validateMinDomains(t *testing.T) {
```
No need to test this function separately.

All the functionality is already covered in `TestValidateTopologySpreadConstraints`

alculquicondor commented:
Thanks, ready for squash

/lgtm

alculquicondor commented:
/hold
for implementation :)

alculquicondor commented:
The piping of the value to kube-scheduler code is in this PR, but the actual logic implementation is in #108362

Not how I would have chosen to split the PRs, but too late now.

sanposhiho commented:
Rebased to resolve conflict.

sanposhiho commented:
@alculquicondor is it ok to unhold as well?

alculquicondor commented:
Only after #107674 merges
